src/ECL/ECL.hs

```diff
@@ -99,56 +142,79 @@ missingSingle (Record bs) (Record bs') =
+missingMultiple env = go
   where
-  go [] _ = []
-  go (x:xs) (y:ys) = map (: xs) missed ++ fmap (x :) missed'
+  go [] _ = ([], pure True)
+  go (x:xs) (y:ys) = (map (: xs) missed ++ fmap (x :) missed', liftA2 (&&) pr1 pr2)
     where
-    missed = missingSingle x y
-    missed' = go xs ys
+    (missed, pr1) = missingSingle env x y
+    (missed', pr2) = go xs ys
   go _ _ = error "Error in missingMultiple: invalid length of argument binders."
 
 -- |
--- `missingCases` applies `missingMultiple` to an alternative
+-- `missingCases` applies `missingMultiple` to an alternative.
 --
-missingCases :: (Eq lit) => Binders lit -> Alternative lit -> Uncovered lit
-missingCases unc = missingMultiple unc . fst
+missingCases :: (Eq lit) => Environment -> Binders lit -> Alternative lit -> ([Binders lit], Maybe Bool)
+missingCases env unc = missingMultiple env unc . fst
 
 -- |
--- `missingAlternative` is `missingCases` with guard handling
+-- `missingAlternative` is `missingCases` with guard handling.
 --
-missingAlternative :: (Eq lit) => Alternative lit -> Binders lit -> [Binders lit]
-missingAlternative alt unc
-  | isExhaustiveGuard $ snd alt = getUncovered mcases
-  | otherwise = [unc]
+missingAlternative :: (Eq lit) => Environment -> Alternative lit -> Binders lit -> ([Binders lit], Maybe Bool)
+missingAlternative env alt unc
+  | isExhaustiveGuard $ snd alt = mcases
+  | otherwise = ([unc], snd mcases)
   where
-  mcases = missingCases unc alt
+  mcases = missingCases env unc alt
 
   isExhaustiveGuard :: Maybe Guard -> Bool
   isExhaustiveGuard (Just Opaque) = False
   isExhaustiveGuard _ = True
 
 -- |
--- Given a list of alternatives, `check` generates the proper set of uncovered cases
+-- Given a list of alternatives, `check'` generates the proper set of uncovered cases.
+--
+check' :: (Eq lit) => Environment -> [Alternative lit] -> Check lit
+check' env cas = applyUncovered nub . foldl' step initial $ cas
+  where
+  initial = makeCheck [initialize $ length . fst . head $ cas] $ (pure True, [])
+
+  step :: (Eq lit) => Check lit -> Alternative lit -> Check lit
+  step ch ca =
+    let (missed, pr) = unzip $ map (missingAlternative env ca) $ getUncovered ch
+        (mb, redundant) = getRedundant ch
+        cond = liftA2 (&&) (or <$> sequenceA pr) mb
+    in Check
+      { getUncovered = if fromMaybe True cond then concat missed else getUncovered ch
+      , getRedundant = (cond, if fromMaybe True cond then redundant else fst ca : redundant) }
+    where
+    sequenceA = foldr (liftA2 (:)) (pure [])
+
+-- |
+-- Given two translation functions (between the desired type and `Binder`) an environment 
+-- and a list of alternatives, `check` generates the proper set of uncovered cases.
 --
-check :: (Eq lit) => [Alternative lit] -> [Binders lit]
-check cas = getUncovered . applyUncovered nub . foldl' step (Uncovered [initial]) $ cas
+check :: (Eq a, Eq lit) =>
```

Comment:
I would let the user convert their types to `Binder` and back. Doing the function composition for them just limits this function's usefulness.